lib/src/request_processor.cc

```diff
@@ -412,7 +391,312 @@ void RequestProcessor::processNonBlockingRequest(const ActionRequest& request) {
+// |   yes   |    no    |   yes   |     -    |     -     |  SUCCESS / FAILURE  |
+// |         |          |         |          |           | + stdout & stderr   |
+// |         |          |         |          |           | + metadata update   |
+// |+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++|
+// |   yes   |    yes   |         |     -    |     -     |  SUCCESS / FAILURE  |
+// |         |          |         |          |           | + stdout & stderr   |
+// |+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++|
+//
+
+void RequestProcessor::processStatusRequest(const ActionRequest& request)
+{
+    ActionResponse status_response { ModuleType::Internal, request };
+    lth_jc::JsonContainer status_results {};
+    const auto& AS = ACTION_STATUS_NAMES;
+    auto t_id = request.params().get<std::string>("transaction_id");
+    status_results.set<std::string>("transaction_id", t_id);
+    status_results.set<std::string>("status", AS.at(ActionStatus::Unknown));
+
+    if (!storage_ptr_->find(t_id)) {
+        LOG_DEBUG("Found no results for the %1%", request.prettyLabel());
+        status_response.setValidResultsAndEnd(std::move(status_results),
+                                              "found no results directory");
+        connector_ptr_->sendBlockingResponse(status_response, request);
+        return;
+    }
+
+    // There's a results directory for the requested transaction!
+
+    // Get PID information before metadata, to avoid a race with the
+    // nonBlockingActionTask thread when accessing the metadata file,
+    // due to the elapsed time between the end of the external module
+    // process and the metadata update (note that we use a lock based
+    // on a named mutex to access the medata file in a safe way)
+
+    bool running_by_pid { false };
+    bool not_running_by_pid { false };
+
+    if (!storage_ptr_->pidFileExists(t_id)) {
+        LOG_DEBUG("PID file for transaction %1% task does not exist", t_id);
+    } else {
+        try {
+            auto pid = storage_ptr_->getPID(t_id);
+
+            // NOTE(ale): processExists() does not throw
+            if (Util::processExists(pid)) {
+                running_by_pid = true;
+            } else {
+                not_running_by_pid = true;
+                bool cached { false };
+                {
+                    ResultsMutex::LockGuard a_l { ResultsMutex::Instance().access_mtx };
+                    cached = ResultsMutex::Instance().exists(t_id);
+                }
+                if (cached)
+                    // The process doe not exist anymore, but its
+                    // mutex is still cached - wait a bit before
+                    // reading the metadata to allow the non-blocking
+                    // to lock its mutex and update the metadata
+                    pcp_util::this_thread::sleep_for(
+                        pcp_util::chrono::milliseconds(METADATA_RACE_MS));
```

Comment:
It would be great to know which thread is handling the action and just wait for it to finish.